R/src/feather-write.cpp

```diff
@@ -106,12 +106,12 @@ PrimitiveArray factorCodesToPrimitiveArray(SEXP x) {
 
   auto values_buffer = std::make_shared<OwnedMutableBuffer>();
   stopOnFailure(values_buffer->Resize(n * sizeof(int32_t)));
-  memset(values_buffer->mutable_data(), 0, n);
+  util::fill_buffer(values_buffer->mutable_data(), 0, n);
   auto values = reinterpret_cast<int32_t*>(values_buffer->mutable_data());
 
   for (int i = 0; i < n; ++i) {
-    // Results of NA values don't matter as the value slots are undefined
```

Comment:
There's a validity bitmap (1 for valid, 0 for NA). This should be used on reconstructing the factor codes (setting whatever is the NA code in the slot)